File: slack-reader/src/test/java/dev/number6/slackreader/core/SnsServiceTest.java
package dev.number6.slackreader.core;

import dev.number6.slackreader.SnsService;
import dev.number6.slackreader.model.WorkspaceMessages;
import dev.number6.slackreader.port.NotificationPort;
import dev.number6.slackreader.generate.SlackReaderRDG;
import dev.number6.message.ChannelMessages;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import uk.org.fyodor.range.Range;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@ExtendWith(MockitoExtension.class)
class SnsServiceTest {

    @Mock
    private NotificationPort notificationPort;

    private SnsService testee;

    @BeforeEach
    void setup() {

        testee = new SnsService(notificationPort);
    }

    @Test
    void sendMessagesChannelToSns() {

        WorkspaceMessages workspaceMessages = new WorkspaceMessages(LocalDate.now());
        String channelName = SlackReaderRDG.string().next();
        List<String> messages = SlackReaderRDG.list(SlackReaderRDG.string(), Range.closed(10, 30)).next();

        workspaceMessages.add(channelName, messages);
        testee.broadcastWorkspaceMessagesForActiveChannels(workspaceMessages);

        ArgumentCaptor<ChannelMessages> captor = ArgumentCaptor.forClass(ChannelMessages.class);
        verify(notificationPort, times(1)).broadcast(captor.capture());

        assertThat(captor.getValue().getComprehensionDate()).isEqualTo(LocalDate.now());
        assertThat(captor.getValue().getChannelName()).isEqualTo(channelName);
        assertThat(captor.getValue().getMessages()).hasSameElementsAs(messages);
    }

    @Test
    void sendMessagesForActiveChannelToSns() {

        WorkspaceMessages workspaceMessages = SlackReaderRDG.workspaceMessages(Range.closed(5, 20)).next();

        String channelName = SlackReaderRDG.string().next();
        workspaceMessages.add(channelName, new ArrayList<>());

        testee.broadcastWorkspaceMessagesForActiveChannels(workspaceMessages);

        ArgumentCaptor<ChannelMessages> captor = ArgumentCaptor.forClass(ChannelMessages.class);
        verify(notificationPort, times(workspaceMessages.getActiveChannelNames().size())).broadcast(captor.capture());

        assertThat(captor.getAllValues()).hasSize(workspaceMessages.getActiveChannelNames().size());
        for (ChannelMessages channelMessages : captor.getAllValues()) {
            assertThat(channelMessages.getComprehensionDate()).isEqualTo(LocalDate.now());
            assertThat(channelMessages.getChannelName()).isIn(workspaceMessages.getActiveChannelNames());
            assertThat(channelMessages.getMessages()).containsExactlyInAnyOrderElementsOf(workspaceMessages.getMessagesForChannel(channelMessages.getChannelName()));
        }
    }
}

File: slack-reader/src/test/kotlin/dev/number6/slackreader/core/SnsServiceTest.kt
package dev.number6.slackreader.core

import assertk.assertThat
import assertk.assertions.containsOnly
import assertk.assertions.hasSameSizeAs
import assertk.assertions.isEqualTo
import assertk.assertions.isIn
import dev.number6.message.ChannelMessages
import dev.number6.slackreader.SnsService
import dev.number6.slackreader.generate.SlackReaderRDG
import dev.number6.slackreader.model.WorkspaceMessages
import dev.number6.slackreader.port.NotificationPort
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import uk.org.fyodor.generators.RDG
import uk.org.fyodor.range.Range
import java.time.LocalDate
import java.util.*

@ExtendWith(MockKExtension::class)
internal class SnsServiceTest {

    private val notificationPort: NotificationPort = mockk(relaxUnitFun = true)
    private val testee = SnsService(notificationPort)

    @Test
    fun sendMessagesChannelToSns() {
        val workspaceMessages = WorkspaceMessages(LocalDate.now())
        val expectedChannelName = RDG.string().next()
        val expectedMessages = RDG.list(RDG.string(), Range.closed(10, 30)).next()
        workspaceMessages.add(expectedChannelName, expectedMessages)
        testee.broadcastWorkspaceMessagesForActiveChannels(workspaceMessages)
        val slot = slot<ChannelMessages>()
        verify(exactly = 1) { notificationPort.broadcast(capture(slot)) }

        with(slot.captured) {
            assertThat(comprehensionDate).isEqualTo(LocalDate.now())
            assertThat(channelName).isEqualTo(expectedChannelName)
            assertThat(messages).containsOnly(*expectedMessages.toTypedArray())
        }
    }

    @Test
    fun sendMessagesForActiveChannelToSns() {
        val workspaceMessages = SlackReaderRDG.workspaceMessages(Range.closed(5, 20)).next()
        val channelName = RDG.string().next()
        workspaceMessages.add(channelName, ArrayList())
        testee.broadcastWorkspaceMessagesForActiveChannels(workspaceMessages)
        val broadcastMessages = mutableListOf<ChannelMessages>()
        verify(exactly = workspaceMessages.getActiveChannelNames().size) { notificationPort.broadcast(capture(broadcastMessages)) }
        assertThat(broadcastMessages).hasSameSizeAs(workspaceMessages.getActiveChannelNames())

        broadcastMessages.forEach {
            assertThat(it.comprehensionDate).isEqualTo(LocalDate.now())
            assertThat(it.channelName).isIn(*workspaceMessages.getActiveChannelNames().toTypedArray())
            assertThat(it.messages).containsOnly(*workspaceMessages.getMessagesForChannel(it.channelName).toTypedArray())
        }
    }
}